src/control/cmd/daos_server/storage.go

```diff
@@ -0,0 +1,135 @@
+}
+
+type storageScanCmd struct {
+	cfgCmd
+	logCmd
+}
+
+func (cmd *storageScanCmd) Execute(args []string) error {
+	srv, err := server.NewControlService(cmd.config)
+	if err != nil {
+		return errors.WithMessage(err, "failed to init ControlService")
+	}
+
+	cmd.log.Info("Scanning locally-attached storage...")
+	scanErrors := make([]error, 0, 2)
+	controllers, err := srv.ScanNVMe()
+	if err != nil {
+		scanErrors = append(scanErrors, err)
+	} else {
+		str, err := common.StructsToString(controllers)
+		if err != nil {
+			scanErrors = append(scanErrors, err)
+		} else {
+			cmd.log.Infof("NVMe: %s", str)
+		}
+	}
+
+	modules, err := srv.ScanSCM()
+	if err != nil {
+		scanErrors = append(scanErrors, err)
+	} else {
+		str, err := common.StructsToString(modules)
+		if err != nil {
+			scanErrors = append(scanErrors, err)
+		} else {
+			cmd.log.Infof("SCM: %s", str)
+		}
+	}
+
+	if len(scanErrors) > 0 {
+		errStr := "scan error(s):\n"
+		for _, err := range scanErrors {
+			errStr += fmt.Sprintf("  %s\n", err.Error())
+		}
+		return errors.New(errStr)
+	}
+
+	return nil
+}
+
+type storagePrepNvmeCmd struct {
+	cfgCmd
+	PCIWhiteList string `short:"w" long:"pci-whitelist" description:"Whitespace separated list of PCI devices (by address) to be unbound from Kernel driver and used with SPDK (default is all PCI devices)."`
+	NrHugepages  int    `short:"p" long:"hugepages" description:"Number of hugepages to allocate (in MB) for use by SPDK (default 1024)"`
+	TargetUser   string `short:"u" long:"target-user" description:"User that will own hugepage mountpoint directory and vfio groups."`
+	Reset        bool   `short:"r" long:"reset" description:"Reset SPDK returning devices to kernel modules"`
+}
+
+func (cmd *storagePrepNvmeCmd) Execute(args []string) error {
+	ok, usr := common.CheckSudo()
+	if !ok {
+		return errors.New("subcommand must be run as root or sudo")
+	}
+
+	// falls back to sudoer or root if TargetUser is unspecified
+	tUsr := usr
+	if cmd.TargetUser != "" {
+		tUsr = cmd.TargetUser
+	}
+
+	srv, err := server.NewControlService(cmd.config)
+	if err != nil {
+		return errors.WithMessage(err, "initialising ControlService")
+	}
+
+	return srv.PrepNvme(server.PrepNvmeRequest{
+		HugePageCount: cmd.NrHugepages,
+		TargetUser:    tUsr,
+		PCIWhitelist:  cmd.PCIWhiteList,
+		ResetOnly:     cmd.Reset,
+	})
+}
+
+type storagePrepScmCmd struct {
+	cfgCmd
+	logCmd
+	Reset bool `short:"r" long:"reset" description:"Reset modules to memory mode after removing namespaces"`
+}
+
+func (cmd *storagePrepScmCmd) Execute(args []string) error {
+	ok, _ := common.CheckSudo()
+	if !ok {
+		return errors.New("subcommand must be run as root or sudo")
+	}
+
+	srv, err := server.NewControlService(cmd.config)
+	if err != nil {
+		return errors.WithMessage(err, "initialising ControlService")
+	}
+
+	rebootStr, devices, err := srv.PrepScm(server.PrepScmRequest{
+		Reset: cmd.Reset,
```

Comment:
this kind of changing behaviour should really be in its own PR IMO

Reply:
When these commands lived in the server package, they could use private struct members in the previously-unexported control service. I had a choice here between making those struct members public or expanding the control service API to hide some of the complexity. As I don't think callers really should know about the details I chose the latter approach.